docs/RFCS/sql_batching.md

```diff
@@ -0,0 +1,499 @@
+discussion on indirect statement interaction). Using our previous notation, each statement
+will have an empty write-set (W<sub>1</sub> = W<sub>2</sub> = Ø), so all both intersections
+in the statement independence rules will produce the empty set.
+
+Perhaps more interestingly, we can also reason that two statements that perform mutations
+(INSERT, UPDATE, DELETE, etc.) but that touch disjoint sets of data are also independent
+of one another. It is important to note here that the set of data that a statement _touches_
+includes any that the statement mutates and any that the statement reads while performing
+validation and filtering.
+
+Taking this a step further, we can use our statement independence rules to find that two
+mutating statements can be independent even if their touched data sets overlap, as long
+as only their read-sets overlap. A practical example of this result is two statements that
+insert rows into two different tables, but each validates a foreign key in the same third
+table. These two statements are independent because only their read-sets overlap.
+
+Finally, we can reason that if there is overlap between the touched sets of data between
+any two statements where at least one of the two performs mutation in the overlapping region,
+then the statements are not independent.
+
+Examples:
+
+```
+SELECT * FROM a
+SELECT * FROM b
+SELECT * FROM a
+```
+All three statements are **independent** because they have empty write-sets.
+
+```
+INSERT INTO a VALUES (1)
+INSERT INTO b VALUES (1)
+```
+Statements are **independent** because they have disjoint write-sets and empty read-sets.
+
+```
+UPDATE a SET x = 1 WHERE y = false
+UPDATE a SET x = 2 WHERE y = true
+```
+Statements are **independent** because they have disjoint write-sets and their read-sets
+do not overlap with the other's write-sets. (note that parallelizing this example will not
+actually be supported by the initial proposal, see [Dependency Granularity](#dependency-granularity))
+
+```
+UPDATE a SET b = 2 WHERE y = 1
+UPDATE a SET b = 3 WHERE y = 1
+```
+Statements are **dependent** because write-sets overlap.
+
+```
+UPDATE a SET y = true  WHERE y = false
+UPDATE a SET y = false WHERE y = true
+```
+Statements are **dependent** because while write-sets are disjoint, the read-sets for
+each statement overlap with the other statement's write-set.
+
+```
+UPDATE a SET x = 1    WHERE y = false
+UPDATE a SET y = true WHERE x = 1
+```
+Statements are **dependent** because while write-sets are disjoint, the read-sets for
+each statement overlap with the other statement's write-set.
+
+```
+INSERT INTO a VALUES (1)
+INSERT INTO a VALUES (2)
+```
+Statements are **independent** because they write to different sets of data. However,
+these will not be considered independent by this proposal because writes to the same
+table may implicitly mutate state on the entire table (e.g. a `SERIAL` counter),
```

Comment:
And wouldn't that counter just be subsumed into the wire set and thus force a conflict? Once that's done (though of course easy to miss side effects) couldn't we speed up bulk inserts that way?